### PR TITLE
RavenDB-11880 add history of index Definition ip address and creation…

### PR DESCRIPTION
**Pull request by talweiss1982**
… time

### CODE REVIEW

**Review by ayende — COMMENTED** [review]
How can we access this data? Do we have a way to get the full db record?

**Comment by talweiss1982** [comment]
We have access to the DatabaseRecord see:
<server_url>/studio/index.html#databases/settings/databaseRecord?&database=<db_name>

File: src/Raven.Server/Documents/Indexes/IndexStore.cs [hunk 1]
```diff
@@ -430,14 +430,14 @@ public Index GetIndex(string name)
...
-        public async Task<Index> CreateIndex(IndexDefinition definition)
+        public async Task<Index> CreateIndex(IndexDefinition definition, string source = null)
         {
             if (definition == null)
                 throw new ArgumentNullException(nameof(definition));
 
             ValidateStaticIndex(definition);
 
-            var command = new PutIndexCommand(definition, _documentDatabase.Name);
+            var command = new PutIndexCommand(definition, _documentDatabase.Name, source, DateTime.UtcNow);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Indexes/IndexStore.cs]: use `Database.Time.GetUtc()`

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 2]
```diff
@@ -134,7 +134,7 @@ public void WriteIndex(IndexDefinitionBase indexDefinition, IndexType indexType)
             {
                 if (_batch != null)
                 {
-                    _batch.AddIndex(indexDefinition);
+                    _batch.AddIndex(indexDefinition, _source, DateTime.UtcNow);
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: use `Database.Time.GetUtc()`

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 3]
```diff
@@ -145,13 +145,15 @@ public void WriteIndex(IndexDefinition indexDefinition)
...
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: use `Database.Time.GetUtc()`

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 4]
```diff
@@ -145,13 +145,15 @@ public void WriteIndex(IndexDefinition indexDefinition)
...
                     return;
                 }
 
-                AsyncHelpers.RunSync(() => _database.IndexStore.CreateIndex(indexDefinition));
+                AsyncHelpers.RunSync(() => _database.IndexStore.CreateIndex(indexDefinition, _source));
             }
 
+            private const string _source = "From smuggler";
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: just `Smuggler`

**Review by ppekrol — COMMENTED** [review]
We need to create an endpoint `/databases/*/indexes/history?name=<name>` 
When it will be ready we can add this to Studio (create a separate issue for this)

**Comment by talweiss1982** [comment]
done

File: src/Raven.Server/Documents/Handlers/IndexHandler.cs [hunk 5]
```diff
@@ -102,6 +103,47 @@ public Task Source()
...
 
+        [RavenAction("/databases/*/indexes/history", "GET", AuthorizationStatus.ValidUser)]
+        public Task GetIndexHistory()
+        {
+            var name = GetQueryStringValueAndAssertIfSingleAndNotEmpty("name");
+            var record = Database.ReadDatabaseRecord();
+            List<IndexHistoryEntry> history;
+            if (record.IndexesHistory.TryGetValue(name, out history) == false || history == null || history.Count  == 0) 
+            {
+                throw new InvalidOperationException("Could not find an history for index: " + name);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/IndexHandler.cs]: Let's return 404 here to match behavior from other endpoints.

File: src/Raven.Server/Documents/Handlers/IndexHandler.cs [hunk 6]
```diff
@@ -102,6 +103,47 @@ public Task Source()
             return Task.CompletedTask;
         }
 
+        [RavenAction("/databases/*/indexes/history", "GET", AuthorizationStatus.ValidUser)]
+        public Task GetIndexHistory()
+        {
+            var name = GetQueryStringValueAndAssertIfSingleAndNotEmpty("name");
+            var record = Database.ReadDatabaseRecord();
+            List<IndexHistoryEntry> history;
+            if (record.IndexesHistory.TryGetValue(name, out history) == false || history == null || history.Count  == 0)
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/IndexHandler.cs]: I think we can get NRE when accessing `IndexesHistory` ?

File: src/Raven.Server/Documents/Handlers/IndexHandler.cs (hunk 5 above)
Comment (ppekrol) [reply src/Raven.Server/Documents/Handlers/IndexHandler.cs]: Let's only return 404 when index is not found. When history is null or count is 0 just return empty array.

**Review by ppekrol — APPROVED** [review]

**Pull request merged**